File: backend/app/services/file_handler.py
"""文件处理服务 - 迁移自原项目 src/core/file_handler.py"""
from pathlib import Path
from typing import Optional

import chardet


DEFAULT_ENCODING = "utf-8"
ENCODE_DETECT_SAMPLE_SIZE = 4096
ENCODING_CONFIDENCE_THRESHOLD = 0.7


class FileOperationError(Exception):
    def __init__(self, message: str, file_path: str | None = None):
        self.file_path = file_path
        super().__init__(message)


class EncodingDetectionError(Exception):
    pass


class FileHandler:
    """文件处理器类 - 从原项目迁移，保持业务逻辑不变"""

    @staticmethod
    def detect_encoding(file_path: Path) -> str:
        try:
            with open(file_path, 'rb') as file:
                raw_data = file.read(ENCODE_DETECT_SAMPLE_SIZE)
                result = chardet.detect(raw_data)
                encoding = result.get('encoding')
                confidence = result.get('confidence', 0)
                if confidence < ENCODING_CONFIDENCE_THRESHOLD or not encoding:
                    return DEFAULT_ENCODING
                if encoding and encoding.upper() == 'GB2312':
                    encoding = 'gb18030'
                return encoding
        except IOError as e:
            raise FileOperationError(f"无法读取文件进行编码检测: {e}", str(file_path)) from e
        except Exception as e:
            raise EncodingDetectionError(f"编码检测失败: {e}") from e

    @staticmethod
    def read_file(file_path: Path, encoding: Optional[str] = None) -> str:
        try:
            if encoding is None:
                encoding = FileHandler.detect_encoding(file_path)
            with open(file_path, 'r', encoding=encoding) as file:
                return file.read()
        except UnicodeDecodeError:
            try:
                with open(file_path, 'r', encoding='gb18030') as file:
                    return file.read()
            except UnicodeDecodeError as e:
                raise FileOperationError(f"文件解码失败: {e}", str(file_path)) from e
        except IOError as e:
            raise FileOperationError(f"无法读取文件: {e}", str(file_path)) from e
        except Exception as e:
            raise FileOperationError(f"读取文件时发生未知错误: {e}", str(file_path)) from e

    @staticmethod
    def save_file(file_path: Path, content: str, encoding: str = DEFAULT_ENCODING) -> None:
        try:
            file_path.parent.mkdir(parents=True, exist_ok=True)
            with open(file_path, 'w', encoding=encoding) as file:
                file.write(content)
        except IOError as e:
            raise FileOperationError(f"无法保存文件: {e}", str(file_path)) from e
        except Exception as e:
            raise FileOperationError(f"保存文件时发生未知错误: {e}", str(file_path)) from e

File: frontend/lib/core/file_handler.dart
/// 文件处理核心模块
///
/// 提供文件编码检测、读取、保存等功能。
/// 功能等价于 Python 版 src/core/file_handler.py
///
/// 注意：Dart/Flutter 没有直接的 chardet 等价物，
/// 编码检测使用 dart:convert 的 fallback 策略。

import 'dart:convert';
import 'dart:io';

class FileHandler {
  static const String defaultEncoding = 'utf-8';
  static const int encodeDetectSampleSize = 4096;
  static const double encodingConfidenceThreshold = 0.7;

  /// 检测文件编码
  ///
  /// Dart 没有 chardet，使用简化策略：
  /// 1. 尝试 UTF-8 解码采样字节
  /// 2. 如果失败，尝试 GBK (gb18030 兼容)
  /// 3. 都失败则返回默认 UTF-8
  static String detectEncoding(String filePath) {
    try {
      final file = File(filePath);
      final rawBytes = file.readAsBytesSync().take(encodeDetectSampleSize).toList();

      // 尝试 UTF-8
      try {
        utf8.decode(rawBytes);
        return 'utf-8';
      } catch (_) {
        // 不是有效 UTF-8
      }

      // 尝试 GBK (gb18030 兼容)
      try {
        gbk.decode(rawBytes);
        return 'gb18030';
      } catch (_) {
        // 不是有效 GBK
      }

      return defaultEncoding;
    } catch (e) {
      throw Exception('编码检测失败: $e');
    }
  }

  /// 读取文件内容
  ///
  /// 如果 encoding 为 null 则自动检测
  static String readFile(String filePath, {String? encoding}) {
    encoding ??= detectEncoding(filePath);

    try {
      final content = File(filePath).readAsStringSync(encoding: Encoding.getByName(encoding));
      return content;
    } on FormatException catch (e) {
      // 解码失败，尝试 gb18030
      try {
        return File(filePath).readAsStringSync(encoding: Encoding.getByName('gb18030'));
      } catch (e2) {
        throw Exception('文件解码失败，尝试编码 $encoding 和 gb18030 均失败: $e2');
      }
    } catch (e) {
      throw Exception('无法读取文件: $e');
    }
  }

  /// 保存文件内容
  static void saveFile(String filePath, String content, {String encoding = 'utf-8'}) {
    try {
      final file = File(filePath);
      final dir = file.parent;
      if (!dir.existsSync()) {
        dir.createSync(recursive: true);
      }
      file.writeAsStringSync(content, encoding: Encoding.getByName(encoding));
    } catch (e) {
      throw Exception('无法保存文件: $e');
    }
  }
}
